### ISSUE TITLE
[C++20] [Modules] Incorrect ODR detection for unresolved using declaration

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#52713]
We met a curious problem when practicing the C++20 module. Here is a reduced example:

```C++
// foo.h
template <class Base>
struct bar {
using Ty = int;
};
template <class T>
struct foo : bar<T> {
using typename bar<T>::Ty;
void baz(Ty);
};

// X.cppm
module;
#include "foo.h"
export module X;

// Y.cppm
module;
#include "foo.h"
export module Y;
import X;
```

And the compiler would emit an incorrect ODR violation:
```
foo.h:8:6: error: 'foo::baz' from module 'X.<global>' is not present in definition of 'foo<T>' provided earlier
void baz(Ty);
     ^
foo.h:8:6: note: declaration of 'baz' does not match
void baz(Ty);
     ^
```

From the diagnostic message, we could know it isn't correct clearly. Since even the locations are the same!

The curious point here is that if we write `baz` as `void baz(typename bar<T>::Ty);` or `using Ty =  typename bar<T>::Ty`, then the compiler would behave correctly.

**Comment by ChuanqiXu9** [comment]
Tried to fix this in https://reviews.llvm.org/D115792

**Comment by ChuanqiXu9** [comment]
This should be addressed in: https://github.com/llvm/llvm-project/commit/c50a4b3f97497c27ad62797080b52f501dac5e38
